### DISCUSSION TITLE
Handling of Mixtures

### DISCUSSION THREAD

**Discussion by artsinli** [oemof/tespy#375] [Ideas]
Hello,

I'm a noob in python so I'm not sure if this has been thought of already. Currently I am looking to replace R134a with R513a (a mixture between R1234yf and R134a), and found this 
[https://tespy.readthedocs.io/en/main/modules/fluid_properties.html#tespy-fluid-properties-label](url)
stating that liquid mixtures can't be used. 

CoolProp can use mixtures (http://www.coolprop.org/fluid_properties/Mixtures.html). Couldn't you theoretically have a fluid object instead of an array that has mixture boolean property that calls the appropriate coolprop mixture functions within each module? Let me know what you think.

**Comment by fwitte** [comment]
Dear @artsinli,

thank you for reaching out. That is a very good idea, I think. I tried it with a little test-script in the current dev version of TESPy:

```python
from tespy.networks import Network

nw = Network(
    fluids=["R134a&R1234YF"],
    p_unit="bar",
    T_unit="C"
)

print("Well I'd like to print this, but the execution crashes silently befor this point.")
```

Something just crashes silently in the background (no Error raised anywhere, execution just stops magically). I identified the section of code responsible for that but I do not yet understand why it does that. To fix it, TESPy will needs a way to handle the `&` input on the fluid name separately from other fluids.

While this can be done rather quick, it seems there are two more things we need to do, before we can then make use of the binary mixtures of CoolProp:

- There are some numerical stability issues https://github.com/CoolProp/CoolProp/issues/2000 and
- We'd need to implemenet some sort of root-finding to replace the `AbstractState.update` on enthalpy-pressure and entropy-pressure pairs for such mixtures, see https://github.com/CoolProp/CoolProp/issues/2004

There are ideas to rework the fluid property back-end #372, I would like to try to take this feature into consideration when working on that. It is however not on top of the priority list, so you might need to wait for it a little bit. I am sorry for that :(. Maybe we a group of interested users, who would like to contribute here.

Best

Francesco

**Reply by artsinli** [reply]
```
import CoolProp.CoolProp as CP

# This adds a dummy entry in the library of interaction parameters if the mixture is not already there
CP.apply_simple_mixing_rule('Helium', 'Xenon', 'linear')

# First use the high-level call to PropsSI to get the mass density of a 50/50 mixture:
CP.PropsSI('Dmass','T',300,'P',101325,'Helium[0.5]&Xenon[0.5]')
Out[17]: 2.750733505001462
```

I looked into the mixture states a bit more, it seems like you have to call the mixing rules before and specify the mixture amounts when calling the function. Maybe adding in another keyword into the networks module (ie. mixture=True) would call these functions and error check for amounts specified etc.?

**Reply by artsinli** [reply]
It seems calling the mixing rule before and doing this makes it not crash.
```
CP.apply_simple_mixing_rule('Helium', 'Xenon', 'linear')
    nw = Network(
        fluids=['Helium[0.5]&Xenon[0.5]'],
        p_unit="bar",
        T_unit="C"
    )
    print("Well I'd like to print this, but the execution crashes silently befor this point.")
```
Output:
```
WARNING:root:Could not find the fluid "Helium[0.5]&Xenon[0.5]" in the fluid property database.
Well I'd like to print this, but the execution crashes silently befor this point.
```

**Reply by fwitte** [reply]
Thank you for the hints, I found it:

```python
import CoolProp.CoolProp as CP

from tespy.networks import Network

CP.apply_simple_mixing_rule('R134a', 'R1234YF', 'linear')

nw = Network(
    fluids=["R134a&R1234YF"],
    p_unit="bar",
    T_unit="C"
)

print("Well I'd like to print this, but the execution crashes silently befor this point.")
```
also crashes the script execution. Your example does not crash it because it cannot recognize the fluid (see the warning) in your console output. I have figured out, why it crashed: https://github.com/CoolProp/CoolProp/issues/2205.

So, in case of binary mixtures, mole fractions must be initialized in some way, when creating the network. Passing something like

```python
["R134a[0.3]&R1234YF[0.7]"]
```

to the TESPy Network instance should therefore strip the brackets expressions from the fluid name, create the AbstractState instance and then set the specified mole fractions before retrieving the boundaries of the fluid property implementation (minimum and maximum temperature and pressure of the fluid in CoolProp).
